### DISCUSSION TITLE
Axi4-Stream for IP interfacing

### DISCUSSION THREAD

**Discussion by kleinai** [SpinalHDL/SpinalHDL#657] [Ideas]
I started looking into an AXI Streaming framework. I use Vivado IP a lot and most of those use AXI Stream compliant interfaces.

The plan was to provide a mapping between standard Spinal streams and AXI Streams for external data movement and to blackbox IP. Maybe some additional components for pre-formatting incoming/out-going AXI Streams (resizing, strobe/keep, etc..) as defined by the AMBA spec.

Right now I'm only interested into fitting stream data into N bytes of AXI Stream data or popping data back out.

See:
https://github.com/kleinai/SpinalHDL/commit/9b8ab7ce9433561c6fece4450d95ac2f96613534

**Comment by Dolu1990** [comment]
Your mapping seems good ^^

I did something in that kind with Bsb : 

https://github.com/SpinalHDL/SpinalHDL/blob/dev/lib/src/main/scala/spinal/lib/bus/bsb/Bsb.scala

Maybe there is some things of interrest in it.

> https://github.com/kleinai/SpinalHDL/commit/9b8ab7ce9433561c6fece4450d95ac2f96613534#diff-36f034cd54c9ca28e6b95bc4af6ac82d2246006898ee1967a5dcfa1577735b24R19

val tid = (config.useId) generate UInt(config.idWidth bit)   :)

def last   =>   def isLast  ? to show it is a function and not a field of the class

**Comment by kleinai** [comment]
I saw that and thought it might work but AXI streams support strobe (mask) and keep which Bsb does not have. Of course positional bytes are a more advanced usage of AXI streams.
There's also special resizing of user bits but I might break that out as to give some user flexibility.

> val tid = (config.useId) generate UInt(config.idWidth bit) :)

Very nice :D

**Comment by kleinai** [comment]
I added some more features. Breaking out IDs, dest, and user bits when converting to a specific type.

https://github.com/kleinai/SpinalHDL/blob/fd0f9b834ed131b071b13adad17b49ab13232c12/lib/src/main/scala/spinal/lib/bus/amba4/axis/Axi4Stream.scala

Is this something you might want to keep around? It's a fairly independent change.

I only wish I could find a way to implicit class `connectFrom` without the need for the `Axi4Stream` class which just serves to handle bundle connections.

**Reply by Dolu1990** [reply]
Hoo, so far i was realy avoiding extending the Stream class, but instead composing with it.

So We could have the following : 

```scala
case class Axi4StreamConfig(...)
case class Axi4StreamPayload(...) extends Bundle

object Axi4Stream{
   def apply(config) : Stream[Axi4StreamPayload] = Stream(Axi4StreamPayload(config))
}

in the lib package object : 
type AxiStream = Stream[Axi4StreamPayload]
```

And then adding implicit methods to Stream[Axi4StreamPayload] via similar construct as : 
https://github.com/SpinalHDL/SpinalHDL/blob/dev/lib/src/main/scala/spinal/lib.scala#L55

In short, composition over ineritance as much as possible.

What do you mean with "Breaking out IDs" ?

**Reply by kleinai** [reply]
Taking your example and some trial and error I've achieved the effect I was looking for. I no longer need the derived class.
There was an issue with the Axi4 spec renamer and the Bsb type after type erasure. An easy fix with a match statement.

> What do you mean with "Breaking out IDs" ?

Something like this when casing from stream bytes to a hard type:
```scala
def toStream[T <: Data](newType: HardType[T], alignment: BitAlignment = LSB): Stream[TupleBundle4[T, UInt, UInt, Bits]] = {
    val that = Stream(TupleBundle4(newType(),
      if (this.tid != null) this.tid else UInt(0 bit),
      if (this.tdest != null) this.tdest else UInt(0 bit),
      if (this.tuser != null) this.tuser else Bits(0 bit)
    ))
```
However I've abandoned that in favor of driving Flow fire streams.

**Comment by xueweiwujxw** [comment]
```scala
package Enc

import spinal.core._
import spinal.lib._

import spinal.sim._
import spinal.core.sim._
import spinal.lib.bus.amba4.axis._

import spinal.core.fiber.Handle

case class streamBitsConvert(
    inaxisConfig: Axi4StreamConfig,
    outaxisConfig: Axi4StreamConfig
) extends Component {
  val io = new Bundle {
    val ins = slave(Axi4Stream(inaxisConfig))
    val outs = master(Axi4Stream(outaxisConfig))
  }
  val a = Axi4StreamWidthAdapter(io.ins, io.outs)
}

object streamBitsConverGen extends App {
  import spinal.lib.sim.StreamReadyRandomizer
  SimConfig.withFstWave
    .doSim(
      streamBitsConvert(
        Axi4StreamConfig(
          dataWidth = 8,
          useKeep = true,
          useLast = true
        ),
        Axi4StreamConfig(
          dataWidth = 1,
          useKeep = true,
          useLast = true
        )
      )
    ) { dut =>
      dut.clockDomain.forkStimulus(10)
      dut.io.ins.payload.keep #= (scala.math.pow(
        2,
        dut.inaxisConfig.dataWidth
      ) - 1).toInt
      dut.io.ins.payload.last #= false
      // StreamReadyRandomizer(dut.io.outs, dut.clockDomain)
      dut.io.outs.ready #= true
      for (idx <- 0 until 100) {
        if (idx > 10) {
          dut.io.ins.valid #= true
          dut.io.ins.payload.data #= idx
        } else {
          dut.io.ins.valid #= false
        }
        dut.clockDomain.waitSampling()
      }
    }
}
```
I found some trouble when using Axi4StreamWidthAdaptor.
I tried to convert a axistream with 32 bits to a axistream with 8 bits. But the input ready signal was pull down when Axi4StreamWidthAdaptor had received 3 clocks, and it would be never rise.
![image](https://user-images.githubusercontent.com/45992640/195081961-174778ef-7db5-442b-a6a2-8a6b66f94c31.png) 
However when I change the input dataWidth to 64, it seems to be right.
![image](https://user-images.githubusercontent.com/45992640/195082153-25f14adb-62bf-4b08-964e-129f1242e872.png) 
SpinalVersion is 1.7.2

**Reply by kleinai** [reply]
I don't see the same behavior with your provided example on dev or 1.7.2.

![image](https://user-images.githubusercontent.com/18756043/195201310-4d99d99c-8300-4993-8ac0-dd1d5852289d.png)

Could you open an issue so we can track it there?

**Reply by xueweiwujxw** [reply]
Yes, the stream of 64 bits works correct, but the stream of 32 bits does not.

**Reply by xueweiwujxw** [reply]
I opened an issue #893

**Reply by Dolu1990** [reply]
Hmm maybe because the seed randomized.
Maybe i should change that behaviour and have the seed fixed by default, and maybe randomized if a given environnement variable is set, that would help repeatability

**Reply by Dolu1990** [reply]
I just pushed that (fixed seed)
